packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,339 @@
+ * @returns {boolean}
+ */
+export const isActionWithinSessionThreshold = (
+  action: UserAction,
+  latestTimestampInSession: number
+): boolean => {
+  return (
+    Math.abs(action.timestamp - latestTimestampInSession) <
+    SESSION_INACTIVITY_THRESHOLD_IN_MS
+  );
+};
+
+/**
+ * Checks if the timestamp is within a specified range.
+ * @param timestamp {number} - The timestamp to check
+ * @param startTimestamp {number} - The start timestamp
+ * @param endTimestamp {number} - The end timestamp
+ * @returns {boolean}
+ */
+const isTimestampInRange = (
+  timestamp: number,
+  startTimestamp: number,
+  endTimestamp: number
+): boolean => timestamp >= startTimestamp && timestamp <= endTimestamp;
+
+/**
+ * Inserts the session in the timeline at the right location.
+ * @param currentSession {UserSession} - The current session
+ * @param ticketCreationDate {number} - The ticket creation date
+ * @param timeline {UserActionTimeline} - The timeline
+ * @returns {void}
+ */
+const insertSessionInTimeline = (
+  session: UserSession,
+  ticketCreationDate: number,
+  timeline: UserActionTimeline
+) => {
+  if (ticketCreationDate < session.start) {
+    timeline.followingSessions.push(session);
+  } else {
+    timeline.precedingSessions.push(session);
+  }
+};
+
+/**
+ * Divides actions into sessions and organizes them in the timeline based on the ticket creation date. This function does the following:
+ * 1) Iterates over the actions and groups them into sessions based on the session inactivity threshold.
+ * 2) Inserts the case creation action in the current session.
+ * 3) Inserts the session in the timeline at the right location.
+ * 4) Returns the timeline with the current session, 2 preceding sessions and 2 following sessions.
+ * @param actions {UserAction[]} - The actions to split
+ * @param ticketCreationDate {string} - The ticket creation date
+ * @returns {UserActionTimeline} - The timeline of user actions (current session and 2 preceding and following sessions)
+ */
+export const splitActionsIntoTimelineSessions = (
+  actions: UserAction[],
+  ticketCreationDate: number
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    if (isActionWithinSessionThreshold(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+      return;
+    }
+    // We check if the case creation is within the current session or within the inactivity threshold after the end of the session.
+    const isCaseCreationIsPartOfCurrentSession = isTimestampInRange(
+      ticketCreationDate,
+      currentSession.start,
+      currentSession.end + SESSION_INACTIVITY_THRESHOLD_IN_MS
```

Comment:
like we are checking for 30 min after the last action of the session, we need to check for 30 min prior to the first action of the session.

Reply:
@erocheleau was saying we did not need to check before only after because this edge case is mostly to cover the case where the case creation takes a while to send and maybe registers right after the last action in a session

Reply:
I think we should stay consistent to our logic and to our definition of what a session is as this could increase the complexity and cause confusion.

For me, if there is a session A and a new a user action that occurred less than 30min before or after this session, this user action should belong to Session A.
And this should apply for everything, ticket creation acton and any other action type. 

cause here it's like us who are creating the edge case by defining a rule then saying that this rule applies for every user action expect for ticket creation actions.

@erocheleau @SimonMilord

Reply:
I agree with you @mmitiche its a weird behaviour to open a case and then do some actions afterwards, but not impossible. I will add that it checks 30 mins after end and 30 mins before as well.